stacks/api/bucket/bucket.controller.ts

```diff
@@ -262,4 +266,73 @@ export class BucketController {
       });
     });
   }
+
+  async clearRemovedFields(
+    bucketDataService: BucketDataService,
+    previousSchema: Bucket,
+    currentSchema: Bucket
+  ) {
+    let removedKeys = this.findRemovedKeys(
+      previousSchema.properties,
+      currentSchema.properties,
+      [],
+      ""
+    );
+    if (removedKeys.length < 1) return;
+
+    let unsetFields = removedKeys.reduce((pre, acc: any, index, array) => {
+      acc = {...pre, [array[index]]: ""};
+      return acc;
+    }, {});
+
+    await bucketDataService.updateMany(previousSchema._id, {}, {$unset: unsetFields});
+  }
+
+  findRemovedKeys(previousSchema: any, currentSchema: any, removedKeys: string[], path: string) {
+    for (const key of Object.keys(previousSchema)) {
+      if (!currentSchema.hasOwnProperty(key)) {
+        removedKeys.push(path ? `${path}.${key}` : key);
+        //we dont need to check child keys of this key anymore
+        continue;
+      }
+      if (this.isObject(previousSchema[key]) && this.isObject(currentSchema[key])) {
+        this.findRemovedKeys(
+          previousSchema[key].properties,
+          currentSchema[key].properties,
+          removedKeys,
+          path ? `${path}.${key}` : key
+        );
+      } else if (this.isArray(previousSchema[key]) && this.isArray(currentSchema[key])) {
+        this.addArrayPattern(
+          previousSchema[key].items,
+          currentSchema[key].items,
+          removedKeys,
+          path ? `${path}.${key}` : key
+        );
+      }
+    }
+    return removedKeys;
+  }
+
+  addArrayPattern(previousSchema: any, currentSchema: any, removedKeys: string[], path: string) {
+    path = `${path}.$[]`;
+    if (this.isArray(previousSchema) && this.isArray(currentSchema)) {
+      this.addArrayPattern(previousSchema.items, currentSchema.items, removedKeys, path);
+    } else if (this.isObject(previousSchema) && this.isObject(currentSchema)) {
+      this.findRemovedKeys(previousSchema.properties, currentSchema.properties, removedKeys, path);
+    }
+  }
+
+  isObject(schema: any) {
+    return schema &&
+      schema.type == "object" &&
+      schema.properties &&
+      Object.keys(schema.properties).length > 0
+      ? true
+      : false;
+  }
+
+  isArray(schema: any) {
+    return schema && schema.type == "array" && schema.items ? true : false;
```

Comment:
Why do we check for the presence of items?

Reply:
The same check like object. If it's not possible to be undefined, I can remove it.

Reply:
I think that checking for the `schema.type` is enough for this logic.